testing/test_pdb.py

```diff
@@ -3210,12 +3224,19 @@ def fn():
 # c
     """.format(
         break_lnum=line_z,
-        filename=__file__,
+        filename=RE_THIS_FILE_CANONICAL,
     ))
 
 
 @pytest.mark.skipif(not hasattr(pdbpp.pdb.Pdb, "error"),
                     reason="no error method")
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason=(
+        "UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in"
+        " position 6998: character maps to <undefined>"
+    ),
+)
```

Comment:
Would be great to figure that one out - where/how does it happen exactly?

Reply:
For reference, the command I'm running for tests is:

```
tox -e py36 --force-dep="pytest @ git+https://github.com/blueyed/pytest@my-master" -- -k test_continue_arg_with_error
```

I only have Py36 available on my windows machine.

----

The full traceback is:

```pytb
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:648: in do_break
    lineno = int(arg)
E   ValueError: invalid literal for int() with base 10: '.foo'

During handling of the above exception, another exception occurred:
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:659: in do_break
    code = func.__code__
E   AttributeError: 'str' object has no attribute '__code__'

During handling of the above exception, another exception occurred:
C:\dev_python\pdbpp\testing\test_pdb.py:3267: in test_continue_arg_with_error
    filename=RE_THIS_FILE_CANONICAL,
C:\dev_python\pdbpp\testing\test_pdb.py:277: in check
    expected, lines = run_func(func, expected)
C:\dev_python\pdbpp\testing\test_pdb.py:259: in run_func
    return expected, runpdb(func, commands)
C:\dev_python\pdbpp\testing\test_pdb.py:193: in runpdb
    func()
C:\dev_python\pdbpp\testing\test_pdb.py:3244: in fn
    x = 1
C:\dev_python\pdbpp\testing\test_pdb.py:3244: in fn
    x = 1
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\bdb.py:48: in trace_dispatch
    return self.dispatch_line(frame)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\bdb.py:66: in dispatch_line
    self.user_line(frame)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:261: in user_line
    self.interaction(frame, None)
c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pdbpp.py:381: in interaction
    return self._interaction(frame, traceback)
c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pdbpp.py:409: in _interaction
    self._cmdloop()
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:321: in _cmdloop
    self.cmdloop()
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\cmd.py:138: in cmdloop
    stop = self.onecmd(line)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:418: in onecmd
    return cmd.Cmd.onecmd(self, line)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\cmd.py:217: in onecmd
    return func(arg)
c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pdbpp.py:1210: in do_continue
    self.do_tbreak(arg)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:706: in do_tbreak
    self.do_break(arg, 1)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:667: in do_break
    (ok, filename, ln) = self.lineinfo(arg)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:740: in lineinfo
    answer = find_function(item, fname)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:100: in find_function
    for lineno, line in enumerate(fp, start=1):
c:\dev_python\pdbpp\.tox\py36\lib\encodings\cp1252.py:23: in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
E   UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 6738: character maps to <undefined>
```

Removing the `set_trace()` line from the dummy function causes the test to not error (but obviously fail):

```diff
def test_continue_arg_with_error():
     def fn():
-        set_trace()
         x = 1
         y = 2
```

Alternatively, removing the `# c.foo` line from the expected regex

```diff
@@ -3252,7 +3253,6 @@ def test_continue_arg_with_error():
 [NUM] > .*fn()
 -> x = 1
    5 frames hidden .*
-# c.foo
 \*\*\* The specified object '.foo' is not a function or was not found along sys.path.
 # c {break_lnum}
 Breakpoint NUM at {filename}:{break_lnum}
```

 will also cause the test to not error (but again, obviously fail because now there's an extra line):

```
----------------------------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------------------------

\[ *[0-9]+\] > .*fn\(\)                                                                |  [36] > c:\dev_python\pdbpp\testing\test_pdb.py(3244)fn()
-> x = 1                                                                               |  -> x = 1
   5 frames hidden .*                                                                  |     5 frames hidden (try 'help hidden_frames')
\*\*\* The specified object '.foo' is not a function or was not found along sys.path.  |  # c 3246    <<<<<
# c 3246                                                                               |  Breakpoint 1 at c:\dev_python\pdbpp\testing\test_pdb.py:3246    <<<<<
Breakpoint  *[0-9]+ at c\:\\dev_python\\pdbpp\\testing\\test_pdb\.py:3246              |  Deleted breakpoint 1 at c:\dev_python\pdbpp\testing\test_pdb.py:3246    <<<<<
Deleted breakpoint  *[0-9]+                                                            |  [36] > c:\dev_python\pdbpp\testing\test_pdb.py(3246)fn()    <<<<<
\[ *[0-9]+\] > .*fn\(\)                                                                |  -> z = 3    <<<<<
-> z = 3                                                                               |     5 frames hidden (try 'help hidden_frames')    <<<<<
   5 frames hidden .*                                                                  |  # c    <<<<<
# c                                                                                    |  <None>    <<<<<
```

Removing both these lines from the expected regex causes the test to pass, but then the test is the same as `test_continue_arg`

```diff
@@ -3252,8 +3253,6 @@ def test_continue_arg_with_error():
 [NUM] > .*fn()
 -> x = 1
    5 frames hidden .*
-# c.foo
-\*\*\* The specified object '.foo' is not a function or was not found along sys.path.
 # c {break_lnum}
 Breakpoint NUM at {filename}:{break_lnum}
 Deleted breakpoint NUM
```

Reply:
This goes through this Python code: https://github.com/python/cpython/blob/1c2e81ed00477255b09a9c3693ddc441979c6c2e/Lib/pdb.py#L668-L676

Adding `print(arg, sys.exc_info())` there shows this in the test for me:
```
# c.foo
.foo (<class 'AttributeError'>, AttributeError("'str' object has no attribute '__code__'"), …
```
`find_function` gets called for me with `'foo' '/home/user/src/pdbpp/testing/test_pdb.py'`.

The code there is the same in 3.6.0: https://github.com/python/cpython/blob/5c4568a05a0a62b5947c55f68f9f2ecfb90a4f12/Lib/pdb.py#L92-L103

So I guess this also fails for you already?
```python
with open("testing/test_pdb.py") as fp:
    for lineno, line in enumerate(fp, start=1):
        print(lineno, repr(line))
```

Using `iconv --from=cp1252 testing/test_pdb.py` chokes at: https://github.com/pdbpp/pdbpp/blob/aabb10b07ac7db1aa5a503cddac9d08dbe58a087/testing/test_pdb.py#L3009

> def test_utf8():
    def fn():
        # Ñ‚ÐµÑiconv: illegal input sequence at position 64082

So for some reason this file is being handled as cp1252, although it is utf8, and has "# -*- coding: utf-8 -*-".

It's an interesting issue, which we should protect against in terms of not crashing in/via https://github.com/dougthor42/pdbpp/blob/aabb10b07ac7db1aa5a503cddac9d08dbe58a087/src/pdbpp.py#L1210.
From a quick glance however it appears to also be a bug in pdb itself in the first place.

I think it's ok to skip the test for now, but mark it with xfail then: https://github.com/pytest-dev/pytest/blob/63c9ad02f4b15d981adf3ec6d34560e94faf0d05/doc/en/skipping.rst#reason-parameter

Reply:
K, I'll xfail it. Discussion moved to #341.

Reply:
I also made the xfail very strict:
+ if a different error is raised, the test suite will fail.
+ if the test passes, the test suite will fail.

This will make it obvious when a code change affects this test.

Both only apply to Windows, of course.

Let me know if you don't want it this way.

Reply:
> K, I'll xfail it. Discussion moved to #341.

Awesome issue (details etc), thanks! :)
Strictness is also cool!